File: cloud/blockstore/public/sdk/python/client/credentials.py
import grpc
import os


_MAX_CERT_FILE_SIZE = 8 * 1024 * 1024
_AUTH_HEADER = "authorization"
_AUTH_METHOD = "Bearer"


class _AuthMetadataPlugin(grpc.AuthMetadataPlugin):

    def __init__(self, auth_token):
        super(grpc.AuthMetadataPlugin, self).__init__()
        self._auth_token = auth_token

    def __call__(self, context, callback):
        callback(((_AUTH_HEADER, _AUTH_METHOD + " " + self._auth_token),), None)


class ClientCredentials(object):

    def __init__(
            self,
            root_certs_file=None,
            cert_file=None,
            cert_private_key_file=None,
            auth_token=None):

        self.root_certs_file = root_certs_file
        self.cert_file = cert_file
        self.cert_private_key_file = cert_private_key_file
        self.auth_token = auth_token

    def get_ssl_channel_credentials(self):
        credentials = grpc.ssl_channel_credentials(
            root_certificates=self._read_file_optional(self.root_certs_file),
            private_key=self._read_file_optional(self.cert_private_key_file),
            certificate_chain=self._read_file_optional(self.cert_file),
        )

        if self.auth_token is not None:
            auth_metadata_plugin = _AuthMetadataPlugin(self.auth_token)
            credentials = grpc.composite_channel_credentials(
                credentials,
                grpc.metadata_call_credentials(auth_metadata_plugin),
            )

        return credentials

    @staticmethod
    def _read_file_optional(filename=None):
        if filename is None:
            return None

        st = os.stat(filename)
        if st.st_size > _MAX_CERT_FILE_SIZE:
            raise Exception("file too large")

        with open(filename, "rb") as f:
            return f.read()

File: cloud/blockstore/public/sdk/go/client/credentials.go
package client

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"os"

	"golang.org/x/oauth2"
	"google.golang.org/grpc"
	grpc_codes "google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/oauth"
	grpc_metadata "google.golang.org/grpc/metadata"
	grpc_status "google.golang.org/grpc/status"
)

////////////////////////////////////////////////////////////////////////////////

const authHeader = "authorization"
const authMethod = "Bearer"

type ClientCredentials struct {
	RootCertsFile      string
	CertFile           string
	CertPrivateKeyFile string
	AuthToken          string
	IAMClient          TokenProvider
}

type TokenProvider interface {
	Token(ctx context.Context) (string, error)
}

func (creds *ClientCredentials) GetSslChannelCredentials() ([]grpc.DialOption, error) {
	cfg := tls.Config{}

	if creds.CertFile != "" {
		cert, err := tls.LoadX509KeyPair(creds.CertFile, creds.CertPrivateKeyFile)
		if err != nil {
			return nil, fmt.Errorf("failed to load client certificate/key: %w", err)
		}

		cfg.Certificates = []tls.Certificate{cert}
	}

	if creds.RootCertsFile != "" {
		pem, err := os.ReadFile(creds.RootCertsFile)
		if err != nil {
			return nil, fmt.Errorf("failed to read root cert file: %w", err)
		}

		pool := x509.NewCertPool()
		ok := pool.AppendCertsFromPEM(pem)
		if !ok {
			return nil, errors.New("failed to parse PEM")
		}

		cfg.RootCAs = pool
	}

	opts := []grpc.DialOption{
		grpc.WithTransportCredentials(credentials.NewTLS(&cfg)),
	}

	if creds.AuthToken != "" {
		token := oauth2.Token{AccessToken: creds.AuthToken}
		opts = append(opts, grpc.WithPerRPCCredentials(oauth.NewOauthAccess(&token)))
	}

	tokenProvider := creds.IAMClient

	if tokenProvider != nil {
		interceptor := func(
			ctx context.Context,
			method string, req,
			reply interface{},
			cc *grpc.ClientConn,
			invoker grpc.UnaryInvoker,
			opts ...grpc.CallOption,
		) error {

			token, err := tokenProvider.Token(ctx)
			if err != nil {
				// NBS-2396
				return grpc_status.Error(grpc_codes.Unauthenticated, err.Error())
			}

			ctx = grpc_metadata.AppendToOutgoingContext(ctx, authHeader, authMethod+" "+token)
			return invoker(ctx, method, req, reply, cc, opts...)
		}

		opts = append(opts, grpc.WithChainUnaryInterceptor(interceptor))
	}

	return opts, nil
}
